instrumentation/opentelemetry-instrumentation-pika/src/opentelemetry/instrumentation/pika/utils.py

```diff
@@ -190,3 +197,75 @@ def _enrich_span(
         span.set_attribute(
             SpanAttributes.NET_PEER_PORT, channel.connection.params.port
         )
+
+
+# pylint:disable=abstract-method
+class ReadyMessagesDequeProxy(ObjectProxy):
+    def __init__(
+        self,
+        wrapped,
+        queue_consumer_generator: _QueueConsumerGeneratorInfo,
+        tracer: Optional[Tracer],
+        consume_hook: HookT = dummy_callback,
+    ):
+        super().__init__(wrapped)
+        self._self_active_token = None
+        self._self_tracer = tracer
+        self._self_consume_hook = consume_hook
+        self._self_queue_consumer_generator = queue_consumer_generator
+
+    def popleft(self, *args, **kwargs):
+        try:
+            # end active context if exists
+            if self._self_active_token:
+                context.detach(self._self_active_token)
+        except Exception as inst_exception:  # pylint: disable=W0703
+            _LOG.exception(inst_exception)
+
+        evt = self.__wrapped__.popleft(*args, **kwargs)
+
+        try:
+            # If a new message was received, create a span and set as active context
+            if isinstance(evt, _ConsumerDeliveryEvt):
+                method = evt.method
+                properties = evt.properties
+                if not properties:
+                    properties = BasicProperties(headers={})
+                if properties.headers is None:
+                    properties.headers = {}
+                ctx = propagate.extract(
+                    properties.headers, getter=_pika_getter
+                )
+                if not ctx:
+                    ctx = context.get_current()
+                message_ctx_token = context.attach(ctx)
+                span = _get_span(
+                    self._self_tracer,
+                    None,
+                    properties,
+                    destination=method.exchange
+                    if method.exchange
+                    else method.routing_key,
+                    span_kind=SpanKind.CONSUMER,
+                    task_name=self._self_queue_consumer_generator.consumer_tag,
+                    operation=MessagingOperationValues.RECEIVE,
+                )
+                context.detach(message_ctx_token)
+                self._self_active_token = context.attach(
+                    trace.set_span_in_context(span)
+                )
+                try:
+                    self._self_consume_hook(span, evt.body, properties)
+                except Exception as hook_exception:  # pylint: disable=W0703
+                    _LOG.exception(hook_exception)
+
+                # We must end the span here, because the next place we can hook
+                # is not the end of the user code, but only when the next message
+                # arrives. we still set this span's context as the active context
+                # so spans created by user code that handles this message will be
+                # children of this one.
+                span.end()
```

Comment:
Is there a need to end span in case of exception as well?

Reply:
good catch!